src/main/java/org/apache/mesos/scheduler/recovery/DefaultRecoveryPlanManager.java

```diff
@@ -203,9 +174,53 @@ public void forceComplete(UUID phaseId, UUID blockId) {
         }
     }
 
+    /**
+     * Updates the recovery plan if necessary.
+     *
+     * 1. Updates existing blocks.
+     * 2. If the task needs recovery, removes any COMPLETED blocks for this task
+     *    (at most one block for a given task can exist).
+     * 3. If the task needs recovery, creates a new PENDING block.
+     *
+     * @param status task status
+     */
     @Override
     public void update(Protos.TaskStatus status) {
-        LOGGER.info("Ignoring status: {}", TextFormat.shortDebugString(status));
+        try {
+            String taskName = TaskUtils.toTaskName(status.getTaskId());
+
+            Optional<Protos.TaskInfo> taskInfo = stateStore.fetchTask(taskName);
+            if (!taskInfo.isPresent()) {
+                LOGGER.error("Failed to fetch TaskInfo for task (name=%s)".format(taskName));
+                return;
+            }
+
+            // 1. Update existing blocks.
+            this.plan.getPhases().forEach(
+                    phase -> phase.getBlocks().forEach(
+                            block -> block.update(status)));
+
+            if (TaskUtils.needsRecovery(status)) {
+                Optional<Block> newBlock = createBlock(taskInfo.get());
+                if (newBlock.isPresent()) {
+                    Collection<Block> blocks = new ArrayList<>();
+
+                    // 2. remove any COMPLETED blocks for this task.
+                    for (Phase phase : this.plan.getPhases()) {
+                        blocks.addAll(phase.getBlocks().stream()
+                                .filter(block -> !(block.isComplete() && block.getName().equals(taskName)))
+                                .collect(Collectors.toList()));
+                    }
+
+                    // 3. Add new PENDING block.
```

Comment:
@gabrielhartmann @mgummelt refactored this class to make use of `update()` method for receiving task updates, and regenerating the plan. Instead of the previous model, where the plan was re-generated only with each offer cycle. These comments are to address bug(s) introduced during that refactoring.

I think @mgummelt's approach is better is terms of consistency with PlanManager design, and it also allows for a single way for all PlanManagers to receives TaskStatuses via Observer pattern.